### PR TITLE
Make recovery APIs cancellable

### PR DESCRIPTION
**Pull request by DaveCTurner**
Relates #55550

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/admin/indices/recovery/TransportRecoveryAction.java [hunk 1]
```diff
@@ -107,4 +111,19 @@ protected ClusterBlockException checkGlobalBlock(ClusterState state, RecoveryReq
     protected ClusterBlockException checkRequestBlock(ClusterState state, RecoveryRequest request, String[] concreteIndices) {
         return state.blocks().indicesBlockedException(ClusterBlockLevel.METADATA_READ, concreteIndices);
     }
+
+    @Nullable // unless running tests that inject extra behaviour
+    private volatile Runnable onShardOperation;
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/admin/indices/recovery/TransportRecoveryAction.java:116]: This is totally stupid, but I just can't see a better way to block these operations for the purposes of this test.

File: test/framework/src/main/java/org/elasticsearch/action/admin/indices/recovery/TransportRecoveryActionHelper.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+/**
+ * Helper methods for {@link TransportRecoveryAction}.
+ */
+public class TransportRecoveryActionHelper {
+
+    /**
+     * Helper method for tests to call {@link TransportRecoveryAction#setOnShardOperation}.
+     */
+    public static void setOnShardOperation(TransportRecoveryAction transportRecoveryAction, Runnable setOnShardOperation) {
```
Comment (original-brownbear) [inline test/framework/src/main/java/org/elasticsearch/action/admin/indices/recovery/TransportRecoveryActionHelper.java:19]: NIT: Maybe just inline this into the test class that actually uses it? Why have a separate method for this?

**Review by original-brownbear — APPROVED** [review]
LGTM

File: test/framework/src/main/java/org/elasticsearch/action/admin/indices/recovery/TransportRecoveryActionHelper.java (hunk 2 above)
Comment (DaveCTurner) [reply test/framework/src/main/java/org/elasticsearch/action/admin/indices/recovery/TransportRecoveryActionHelper.java:19]: I put it here so I could keep `TransportRecoveryAction#setOnShardOperation` package-private (without needing the test itself to be in the same package as the action)

Comment (original-brownbear) [reply test/framework/src/main/java/org/elasticsearch/action/admin/indices/recovery/TransportRecoveryActionHelper.java:19]: ah ok, makes sense :)

**Pull request merged**